Core/Registry/IRegistryQuerier.cs

```diff
@@ -194,7 +194,7 @@ public static bool HasUpdate(this IRegistryQuerier querier, string identifier, G
             {
                 RelationshipResolver resolver = new RelationshipResolver(
                     new CkanModule[] { newest_version },
-                    null,
+                    new CkanModule[] { querier.InstalledModule(identifier).Module },
```

Comment:
This one broke upgrading AD mods, `InstalledModule()` is `null` for those, thus we get an exception which caught below and returns `false`.

The easy fix is to pass no array if it's a DLL. However since DLLs are considered to satisfy _all_ dependencies, it's still possible that an upgrade would fail if the new version violated a versioned dependency.

The cleaner way would be to track installed DLLs separately in `RelationshipResolver`, and give the constructor a new option to tell the DLLs that are about to removed.

Reply:
> This one broke upgrading AD mods

Maybe we need some tests for that. It seems a bit brittle at the moment.

Reply:
By pure luck, adding a `?` also works, because `RelationshipResolver` is conservative with nulls:

```csharp
                    new CkanModule[] { querier.InstalledModule(identifier)?.Module },
```

> The cleaner way would be to track installed DLLs separately in `RelationshipResolver`, and give the constructor a new option to tell the DLLs that are about to removed.

I wonder whether AD mods should have their own `InstalledModule`, possibly with a `CkanModule` with `"kind": "dll"`?

Reply:
> I wonder whether AD mods should have their own `InstalledModule`, possibly with a `CkanModule` with `"kind": "dll"`?

Interesting idea. I'm a bit anxious that things might break if they expect `InstalledModule()`  to only return CKAN-managed modules. And if `InstalledModule()` returns modules that aren't in `InstalledModules`, more things might get confused (or would we also include DLLs in `InstalledModules`?)
And remembering all the bugs we had to fix because we didn't filter out `"kind": "dlc"`...

Reply:
> And remembering all the bugs we had to fix because we didn't filter out `"kind": "dlc"`...

Yes, it's definitely a high risk change. For now, I think I'm going to try writing that test and doing the easy fix.

Reply:
And I'm currently trying out tracking DLLs separately in the RelationshipResolver, so far it brought up some more places where we missed to pass all the data we should've to the resolver.
But since it basically includes the easy fix anyway, I think it's good that you do it as a separate PR, especially if it contains some tests that help catching any problems with more complex changes.

Reply:
> especially if it contains some tests that help catching any problems with more complex changes

Heh, I was just going to do the one simple test of `HasUpdate` with an AD mod, but now I'm looking for a few more...

> * CKAN offers to upgrade modules even if it would violate another module's relationships

Do you remember which relationships these were? Did another installed module have a `depends` with `version`?

Reply:
> Do you remember which relationships these were? Did another installed module have a `depends` with `version`?

Yup, BeyondHome with Parallax: https://github.com/KSP-CKAN/NetKAN/pull/8378
And this one is also interesting since I've read from a lot of people having Parallax installed manually, sometimes because they didn't know of the downgrade functionality. So that got me into the whole AD and upgrade with unsatisfied dependency rabbit hole.